tool/build/renamestr.c

```diff
@@ -0,0 +1,469 @@
+#include "libc/runtime/symbols.internal.h"
+#include "libc/stdio/stdio.h"
+#include "libc/str/str.h"
+#include "libc/sysv/consts/map.h"
+#include "libc/sysv/consts/o.h"
+#include "libc/sysv/consts/prot.h"
+#include "third_party/getopt/getopt.internal.h"
+
+#define VERSION      \
+  "renamestr v0.1\n" \
+  "https://github.com/jart/cosmopolitan\n"
+
+#define MANUAL                                                \
+  " -f FROM -t TO INPUT \n"                                   \
+  "\n"                                                        \
+  "DESCRIPTION\n"                                             \
+  "\n"                                                        \
+  "  in-place string replacement in ELF binary .rodata\n"     \
+  "\n"                                                        \
+  "  this program may be used to replace strings in the\n"    \
+  "  .rodata sections of ELF binaries, in-place.\n"           \
+  "\n"                                                        \
+  "FLAGS\n"                                                   \
+  "\n"                                                        \
+  "  -h         show usage\n"                                 \
+  "\n"                                                        \
+  "  -v         show version\n"                               \
+  "\n"                                                        \
+  "  -f FROM    source string to replace\n"                   \
+  "\n"                                                        \
+  "  -t TO      target string replacement. must be shorter\n" \
+  "             than FROM string for replacement to work\n"   \
+  "\n"                                                        \
+  "  INPUT      ELF binary containing strings to replace\n"   \
+  "\n"
+
+static const char *prog;
+static const char *exepath;
+static Elf64_Shdr *rodata;
+static char *rostart;
+static char *roend;
+static int exefd;
+
+static wontreturn void Die(const char *thing, const char *reason) {
+  tinyprint(2, thing, ": ", reason, "\n", NULL);
+  exit(1);
+}
+
+static wontreturn void DieSys(const char *thing) {
+  perror(thing);
+  exit(1);
+}
+
+static wontreturn void ShowUsage(int rc, int fd) {
+  tinyprint(fd, "USAGE\n\n  ", prog, MANUAL, NULL);
+  exit(rc);
+}
+
+static wontreturn void DieOom(void) {
+  Die("renamestr", "out of memory");
+}
+
+static void *Malloc(size_t n) {
+  void *p;
+  if (!(p = malloc(n)))
+    DieOom();
+  return p;
+}
+
+static void *Realloc(void *p, size_t n) {
+  if (!(p = realloc(p, n)))
+    DieOom();
+  return p;
+}
+
+static void Pwrite(const void *data, size_t size, uint64_t offset) {
+  ssize_t rc;
+  const char *p, *e;
+  for (p = data, e = p + size; p < e; p += (size_t)rc, offset += (size_t)rc) {
+    if ((rc = pwrite(exefd, p, e - p, offset)) == -1) {
+      DieSys(exepath);
+    }
+  }
+}
+
+struct Param {
+  const char *from_string;
+  size_t from_len;
+  const char *to_string;
+  size_t to_len;
+  int count;
+  char *roloc;
```

Comment:
It's a little weird to have multiple purposes in this same data structure. Normally I'd assume a data structure for storing input parameters would be immutable. But this struct is almost acting like a job object.

Reply:
Hmm, each from/to combination needs to go over the whole file -- so we have two strings, the `roloc` refers to current location within the file, and `count` provides number of replacements performed.

Perhaps the `roloc` and `count` parameters can be in a separate struct?

Reply:
Or the `roloc` can be a global, reset for each iteration.